File: src/build-context.ts
import { createHash } from "node:crypto";
import { constants } from "node:fs";
import * as fs from "node:fs/promises";
import { homedir } from "node:os";
import { dirname, isAbsolute, join, relative, resolve } from "node:path";
import ignore from "@balena/dockerignore";
import { inside } from "./config.js";
import { type Context, fail, isCode } from "./model.js";

export async function prepareBuildContext(
  ctx: Context,
  root: string,
  dockerfile: string,
  cache = join(
    homedir(),
    process.platform === "darwin" ? "Library/Caches" : ".cache",
    "contremaitre",
    "builds",
  ),
) {
  root = await fs.realpath(root);
  dockerfile = resolve(root, dockerfile);
  let patterns = "";
  try {
    patterns = await fs.readFile(`${dockerfile}.dockerignore`, "utf8");
  } catch (e) {
    if (!isCode(e, "ENOENT")) throw e;
    try {
      patterns = await fs.readFile(join(root, ".dockerignore"), "utf8");
    } catch (e) {
      if (!isCode(e, "ENOENT")) throw e;
    }
  }
  const matcher = ignore().add(patterns),
    negations = patterns.split("\n").some((p) => p.trim().startsWith("!"));
  await fs.mkdir(cache, { recursive: true, mode: 0o700 });
  const dir = await fs.mkdtemp(join(cache, "context-"));
  const targetRoot = join(dir, "context"),
    definition = join(dir, "definition", "Dockerfile");
  const cleanup = async () => {
    const writable = async (p: string): Promise<void> => {
      const s = await fs.lstat(p);
      if (!s.isDirectory()) return;
      await fs.chmod(p, 0o700);
      for (const n of await fs.readdir(p)) await writable(join(p, n));
    };
    await writable(dir);
    await fs.rm(dir, { recursive: true, force: true });
  };
  try {
    await fs.mkdir(targetRoot, { mode: 0o700 });
    await fs.mkdir(dirname(definition), { mode: 0o700 });
    const docker = await fs.readFile(dockerfile);
    await fs.writeFile(definition, docker, { mode: 0o600 });
    await fs.writeFile(`${definition}.dockerignore`, patterns, { mode: 0o600 });
    const digest = createHash("sha256")
      .update("context-ts-v1\0")
      .update(docker)
      .update("\0")
      .update(patterns);
    let files = 0,
      bytes = 0;
    const walk = async (path: string): Promise<void> => {
      ctx.signal.throwIfAborted();
      const rel = relative(root, path),
        info = await fs.lstat(path),
        target = join(targetRoot, rel);
      const excluded = rel !== "" && matcher.ignores(rel);
      if (info.isDirectory()) {
        if (excluded && !negations) return;
        // Recheck parent confinement before enumeration; never follow directory links.
        if (!inside(root, await fs.realpath(path))) fail(`Build directory escapes context: ${rel}`);
        if (!excluded) await fs.mkdir(target, { recursive: true, mode: 0o700 });
        for (const entry of (await fs.readdir(path)).sort()) await walk(join(path, entry));
        try {
          await fs.stat(target);
        } catch (e) {
          if (isCode(e, "ENOENT")) return;
          throw e;
        }
        digest.update(JSON.stringify([rel, info.mode & 0o777, "directory"]));
        await fs.chmod(target, info.mode & 0o777);
        await fs.utimes(target, info.mtime, info.mtime);
        return;
      }
      if (excluded) return;
      await fs.mkdir(dirname(target), { recursive: true, mode: 0o700 });
      digest.update(JSON.stringify([rel, info.mode & 0o777]));
      if (info.isSymbolicLink()) {
        let link = await fs.readlink(path);
        const dest = resolve(dirname(path), link);
        if (!inside(root, dest)) fail(`Build symlink escapes context: ${rel}`);
        if (isAbsolute(link))
          link = relative(dirname(target), join(targetRoot, relative(root, dest)));
        digest.update(JSON.stringify(link));
        await fs.symlink(link, target);
        return;
      }
      if (!info.isFile()) fail(`Unsupported build context entry: ${rel}`);
      const input = await fs.open(path, constants.O_RDONLY | constants.O_NOFOLLOW);
      let count = 0;
      try {
        const before = await input.stat();
        if (before.ino !== info.ino || before.dev !== info.dev)
          fail(`Build source changed: ${rel}; retry deploy`);
        if (!inside(root, await fs.realpath(path))) fail(`Build file escapes context: ${rel}`);
        const out = await fs.open(target, "wx", 0o600);
        try {
          digest.update(`${info.size}\0`);
          for await (const chunk of input.createReadStream({ autoClose: false })) {
            ctx.signal.throwIfAborted();
            const data = Buffer.from(chunk);
            digest.update(data);
            let offset = 0;
            while (offset < data.length) {
              const r = await out.write(data, offset);
              offset += r.bytesWritten;
            }
            count += data.length;
          }
        } finally {
          await out.close();
        }
        const after = await input.stat();
        if (count !== info.size || after.mtimeMs !== before.mtimeMs || after.size !== before.size)
          fail(`Build source changed: ${rel}; retry deploy`);
      } finally {
        await input.close();
      }
      await fs.chmod(target, info.mode & 0o777);
      await fs.utimes(target, info.mtime, info.mtime);
      files++;
      bytes += count;
    };
    await walk(root);
    return {
      root: targetRoot,
      dockerfile: definition,
      digest: digest.digest("hex"),
      files,
      bytes,
      cleanup,
    };
  } catch (e) {
    await cleanup();
    throw e;
  }
}

File: compat/go/internal/hub/build_context.go
package hub

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"github.com/moby/patternmatcher"
	"github.com/moby/patternmatcher/ignorefile"
)

type BuildRecord struct {
	Digest string `json:"digest"`
	Image  string `json:"image"`
}
type buildContext struct {
	Root, Dockerfile, Digest string
	Files                    int
	Bytes                    int64
	cleanup                  func()
}

// Stage only included entries. In particular, do not let the Apple builder
// repeatedly enumerate excluded dependency trees during its metadata requests.
func prepareBuildContext(ctx context.Context, root, dockerfile string) (_ *buildContext, err error) {
	root, err = filepath.Abs(root)
	if err != nil {
		return nil, err
	}
	if !filepath.IsAbs(dockerfile) {
		dockerfile = filepath.Join(root, dockerfile)
	}
	dockerfile, err = filepath.Abs(dockerfile)
	if err != nil {
		return nil, err
	}
	ignorePath := dockerfile + ".dockerignore"
	if _, e := os.Stat(ignorePath); os.IsNotExist(e) {
		ignorePath = filepath.Join(root, ".dockerignore")
	} else if e != nil {
		return nil, e
	}
	var patterns []string
	if input, e := os.Open(ignorePath); e == nil {
		patterns, err = ignorefile.ReadAll(input)
		input.Close()
		if err != nil {
			return nil, err
		}
	} else if !os.IsNotExist(e) {
		return nil, e
	}
	matcher, err := patternmatcher.New(patterns)
	if err != nil {
		return nil, fmt.Errorf("invalid dockerignore: %w", err)
	}
	// Apple's filesystem sync omits nested entries under macOS temporary
	// directories. Use the user cache, where recursive COPY remains intact.
	cache, err := os.UserCacheDir()
	if err != nil {
		return nil, err
	}
	cache = filepath.Join(cache, "contremaitre", "builds")
	if err = os.MkdirAll(cache, 0700); err != nil {
		return nil, err
	}
	directory, err := os.MkdirTemp(cache, "context-")
	if err != nil {
		return nil, err
	}
	staged := &buildContext{Root: filepath.Join(directory, "context"), Dockerfile: filepath.Join(directory, "definition", "Dockerfile"), cleanup: func() {
		_ = filepath.WalkDir(directory, func(path string, entry fs.DirEntry, e error) error {
			if e == nil && entry.IsDir() {
				_ = os.Chmod(path, 0700)
			}
			return nil
		})
		_ = os.RemoveAll(directory)
	}}
	defer func() {
		if err != nil {
			staged.cleanup()
		}
	}()
	if err = os.MkdirAll(staged.Root, 0700); err != nil {
		return nil, err
	}
	if err = os.MkdirAll(filepath.Dir(staged.Dockerfile), 0700); err != nil {
		return nil, err
	}
	definition, err := os.ReadFile(dockerfile)
	if err != nil {
		return nil, err
	}
	if err = os.WriteFile(staged.Dockerfile, definition, 0600); err != nil {
		return nil, err
	}
	digest := sha256.New()
	fmt.Fprintf(digest, "context-v1\x00%d\x00%s\x00%q\n", len(definition), definition, patterns)
	// os.Root prevents a path swapped for a symlink during copying from escaping
	// the source tree. Symlinks themselves are copied, never traversed.
	type directoryMode struct {
		path string
		info fs.FileInfo
	}
	var directories []directoryMode
	source, err := os.OpenRoot(root)
	if err != nil {
		return nil, err
	}
	defer source.Close()
	err = filepath.WalkDir(root, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if err := ctx.Err(); err != nil {
			return err
		}
		rel, e := filepath.Rel(root, path)
		if e != nil {
			return e
		}
		if rel == "." {
			return nil
		}
		ignored, e := matcher.MatchesOrParentMatches(filepath.ToSlash(rel))
		if e != nil {
			return e
		}
		if entry.IsDir() {
			info, e := entry.Info()
			if e != nil {
				return e
			}
			directories = append(directories, directoryMode{filepath.Join(staged.Root, rel), info})
		}
		if ignored {
			if entry.IsDir() && !matcher.Exclusions() {
				return filepath.SkipDir
			}
			// Negations may re-include descendants. Keep walking in that case instead
			// of guessing which glob prefixes can reach this directory.
			return nil
		}
		info, e := entry.Info()
		if e != nil {
			return e
		}
		target := filepath.Join(staged.Root, rel)
		if e = os.MkdirAll(filepath.Dir(target), 0700); e != nil {
			return e
		}
		fmt.Fprintf(digest, "%q\x00%d\x00", filepath.ToSlash(rel), info.Mode())
		switch {
		case info.IsDir():
			if e = os.MkdirAll(target, 0700); e != nil {
				return e
			}
		case info.Mode()&os.ModeSymlink != 0:
			link, e := source.Readlink(rel)
			if e != nil {
				return e
			}
			resolved := link
			if !filepath.IsAbs(link) {
				resolved = filepath.Join(filepath.Dir(path), link)
			}
			destination, e := filepath.Rel(root, resolved)
			if e != nil || destination == ".." || strings.HasPrefix(destination, ".."+string(filepath.Separator)) {
				return fmt.Errorf("build symlink escapes context: %s", rel)
			}
			// Absolute in-tree links must point at their staged equivalent.
			if filepath.IsAbs(link) {
				link, e = filepath.Rel(filepath.Dir(target), filepath.Join(staged.Root, destination))
				if e != nil {
					return e
				}
			}
			fmt.Fprintf(digest, "%q\x00", link)
			if e = os.Symlink(link, target); e != nil {
				return e
			}
		case info.Mode().IsRegular():
			input, e := source.Open(rel)
			if e != nil {
				return e
			}
			output, e := os.OpenFile(target, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0600)
			if e != nil {
				input.Close()
				return e
			}
			fmt.Fprintf(digest, "%d\x00", info.Size())
			n, copyErr := io.Copy(io.MultiWriter(output, digest), input)
			input.Close()
			closeErr := output.Close()
			if copyErr != nil {
				return copyErr
			}
			if closeErr != nil {
				return closeErr
			}
			if n != info.Size() {
				return fmt.Errorf("build source changed while copying: %s; retry deploy", rel)
			}
			if e = os.Chmod(target, info.Mode().Perm()); e != nil {
				return e
			}
			if e = os.Chtimes(target, info.ModTime(), info.ModTime()); e != nil {
				return e
			}
			staged.Files++
			staged.Bytes += n
		default:
			return fmt.Errorf("unsupported build context entry: %s", rel)
		}
		_, e = digest.Write([]byte{0})
		return e
	})
	if err != nil {
		return nil, err
	}
	for i := len(directories) - 1; i >= 0; i-- {
		dir := directories[i]
		if _, e := os.Stat(dir.path); os.IsNotExist(e) {
			continue
		}
		rel, _ := filepath.Rel(staged.Root, dir.path)
		fmt.Fprintf(digest, "parent:%q:%d\x00", filepath.ToSlash(rel), dir.info.Mode())
		if err = os.Chmod(dir.path, dir.info.Mode().Perm()); err != nil {
			return nil, err
		}
		if err = os.Chtimes(dir.path, dir.info.ModTime(), dir.info.ModTime()); err != nil {
			return nil, err
		}
	}
	// Docker's selected ignore rules must also govern COPY in the staged context.
	// The definition lives outside it, so even an excluded Dockerfile remains usable.
	// An empty Dockerfile-specific file also overrides root ignore rules.
	if err = os.WriteFile(staged.Dockerfile+".dockerignore", []byte(strings.Join(patterns, "\n")+"\n"), 0600); err != nil {
		return nil, err
	}
	rootInfo, err := os.Stat(root)
	if err != nil {
		return nil, err
	}
	if err = os.Chmod(staged.Root, rootInfo.Mode().Perm()); err != nil {
		return nil, err
	}
	fmt.Fprintf(digest, "root-mode:%d", rootInfo.Mode())
	staged.Digest = hex.EncodeToString(digest.Sum(nil))
	return staged, nil
}

func (a Apple) BuildCached(ctx context.Context, root, dockerfile, tag string, previous BuildRecord) (BuildRecord, error) {
	progress(ctx, "Preparing filtered build context")
	staged, err := prepareBuildContext(ctx, root, dockerfile)
	if err != nil {
		return BuildRecord{}, err
	}
	defer staged.cleanup()
	progress(ctx, "Build context: %d files, %.2f MB", staged.Files, float64(staged.Bytes)/1e6)
	if previous.Digest == staged.Digest && previous.Image != "" {
		if _, err := a.output(ctx, "image", "inspect", previous.Image); err == nil {
			progress(ctx, "Reusing unchanged image %s", previous.Image)
			return previous, nil
		}
	}
	args := []string{"build", "--tag", tag, "--file", staged.Dockerfile, "--progress", "plain"}
	// Apple build otherwise recreates a configured builder at its defaults.
	// Inspect only its resource allocation; never emit process environment.
	if data, inspectErr := a.output(ctx, "inspect", "buildkit"); inspectErr == nil {
		var builders []struct {
			Configuration struct {
				Resources struct {
					CPUs   int   `json:"cpus"`
					Memory int64 `json:"memoryInBytes"`
				} `json:"resources"`
			} `json:"configuration"`
		}
		if json.Unmarshal(data, &builders) == nil && len(builders) == 1 {
			r := builders[0].Configuration.Resources
			if r.CPUs > 0 && r.Memory > 0 {
				args = append(args, "--cpus", fmt.Sprint(r.CPUs), "--memory", fmt.Sprint(r.Memory))
				progress(ctx, "Builder: %d CPUs, %d MB RAM", r.CPUs, r.Memory/(1024*1024))
			}
		}
	}
	args = append(args, staged.Root)
	c := a.command(ctx, args...)
	c.Stdout = deploymentOutput(ctx)
	c.Stderr = c.Stdout
	if err = c.Run(); err != nil {
		return BuildRecord{}, err
	}
	return BuildRecord{Digest: staged.Digest, Image: tag}, nil
}
